marlowe-runtime/docs/filtered-chain-sync-spec-1.0.md

```diff
@@ -0,0 +1,375 @@
+# Filtered Chain Sync Protocol
+
+## Introduction
+
+### Purpose
+
+The Filtered Chain Sync Protocol is an application-level protocol for
+efficient, domain-specific traversal of a blockchain. It is a modification of
+the Chain Sync Protocol which allows the client to skip blocks and filter
+information using arbitrary queries. Provided that the server supports the
+specified query, the client is able to dramatically reduce the volume of data
+it must process compared to running the Chain Sync Protocol. This is
+particularly useful for writing the off-chain code for a smart contract, which
+only needs to process a tiny fraction of the full blockchain.
+
+### Requirements
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in RFC 2119.
+
+An implementation is not compliant if it fails to satisfy one or more of the
+MUST or REQUIRED level requirements for the protocols it implements. An
+implementation that satisfies all the MUST or REQUIRED level and all the SHOULD
+level requirements for its protocols is said to be "unconditionally compliant";
+one that satisfies all the MUST level requirements but not all the SHOULD level
+requirements for its protocols is said to be "conditionally compliant."
+
+### Terminology
+
+The following terminology is used to refer to the roles of participands in, and
+objects of, the protocol.
+
+- connection: A transport layer channel over which a client and a server exchange messages.
+- client: The role played by the participant which sends queries and receives chain data.
+- server: The role played by the participant which responds to queries maintains chain data.
+- state: A member of the set of states specified by the state machine of the protocol.
+- message: A member of the set of state transitions specified by the state maching of the protocol.
+- payload: Data associated and sent with a message.
+- query: Part of the payload of a query next message which the client sends to the server to tell it what to send next.
+- error: A protocol level error which is triggered when one party sends an invalid message. Results in the termination of the connection.
+- rejection: An application level error which is triggered when the server rejects a syntactically valid, semantically invalid query sent by the client. Does not result in the termination of the connection.
+- roll forward: A message sent by the server to the client in response to a query.
+- result: Part of the payload of a roll forward message which contains the result of a query.
+- roll backward: A messagge sent by the server to the client when the client's position has been rolled back.
+- tip: The latest chronological point in the blockchain
```

Comment:
Technical, at the node that one is connected to, which may or may not be the latest chronological point globally.